src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -652,15 +660,130 @@ int PeerReplayer::remote_mkdir(const std::string &epath, const struct ceph_statx
   return 0;
 }
 
+PeerReplayer::FileTransferHandlerThreadPool::FileTransferHandlerThreadPool(
+    int _num_threads, int _max_concurrent_large_file,
+    uint64_t _file_size_threshold)
+    : num_threads(_num_threads), max_concurrent_large_file(std::min(
+                                     _num_threads, _max_concurrent_large_file)),
+      file_size_threshold(_file_size_threshold), workers(_num_threads),
+      stop_flag(true) {
+  max_token_sum = file_size_threshold * num_threads * max_concurrent_large_file;
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::activate() {
+  ceph_assert(stop_flag.load());
+  stop_flag = false;
+  for (auto &worker : workers) {
+    worker = std::thread(&FileTransferHandlerThreadPool::run, this);
+  }
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::deactivate() {
+  ceph_assert(!stop_flag.load());
+  stop_flag = true;
+  pick_task.notify_all();
+  give_task.notify_all();
+  exec_task.notify_all();
+  for (auto &worker : workers) {
+    if (worker.joinable()) {
+      worker.join();
+    }
+  }
+  drain_queue();
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::drain_queue() {
+  std::scoped_lock lock(pmtx);
+  while (!task_queue.empty()) {
+    auto& [task, token] = task_queue.front();
+    task(true);
+    current_token_sum -= token;
+    task_queue.pop();
+  }
+}
+
+// K * L + (S - L) * min(file_size, K)
+uint64_t
+PeerReplayer::FileTransferHandlerThreadPool::get_token(uint64_t file_size) {
+  return file_size_threshold * max_concurrent_large_file +
+         std::min(file_size, file_size_threshold) *
+             (num_threads - max_concurrent_large_file);
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::run() {
+  while (true) {
+    std::function<void(bool)> task;
+    int64_t token;
+    {
+      std::unique_lock<std::mutex> lock(pmtx);
+      pick_task.wait(
+          lock, [this] { return (stop_flag.load() || !task_queue.empty()); });
+      if (stop_flag.load()) {
+        return;
+      }
+      std::tie(task, token) = std::move(task_queue.front());
+      task_queue.pop();
+      give_task.notify_one();
+      exec_task.wait(lock, [&token, this] {
```

Comment:
Here is my observation for dual stage checking(I might be nit-picking or overthinking)
Let's say there is a very large task in the front of the queue, if will only be executed if full resource is available. My idea was let other lazy threads to pick some other tasks and let's see whether they can fit in with the current available source.
Advantage:
1. better resource utilization.
2. differentiate 2 state when a thread is waiting for when queue is empty and when it is waiting for the resource available, otherwise even if queue is empty, it will be notified whenever some thread has finished it's execution, but as queue it empty and it will wake up for no-reason and sleep again(I know this is very rare case)

Reply:
@patrakov it is in the code, may be It was a typo of Igor.